### PR TITLE
Add Automated tests for missing coverage in tcms.testcases.views.NewCaseView

### LINKED ISSUE #1614: Automated tests for missing coverage in tcms.testcases.views.NewCaseView
Kiwi TCMS needs additional automated test(s) as mentioned in subject.

**Definition of Done**:
- full coverage for the module or class mentioned in the title. This means both positive/negative scenarios where this makes sense or scenarios which traverse different branches of the code where that makes sense.
- test class/test method names are descriptive and reflect the behavior of the subject under test.  A good scheme to follow is `Test<ClassUnderTest>.test_[who][what][expected_results]`
- CI pass
- All code review passed & PR merged

Hints:
- Coverage report is at https://codecov.io/gh/kiwitcms/Kiwi/tree/master/tcms
- Devel getting started docs: https://kiwitcms.readthedocs.io/en/latest/contribution.html#testing
- Django testing documentation: https://docs.djangoproject.com/en/3.0/topics/testing/

*This issue is part of Kiwi TCMS open source bounty program. For more information see the link(s) in [bounty-program milestone](https://github.com/kiwitcms/Kiwi/milestone/4)*

### PR DESCRIPTION
**Pull request by rish07**
fixes #1614 

- TestNewCase.test_form_invalid tests validity of the forms in NewCaseView on /cases/new view.

### CODE REVIEW

File: tcms/testcases/tests/test_views.py [hunk 1]
```diff
@@ -179,6 +179,21 @@ def setUpTestData(cls):
...
+        data['summary'] = 'Lorem'*54
+        data['default_tester'] = 'NotARealUser'
+        data['product'] = ''
+        data['category'] = ''
+
+        response = self.client.post(self.new_case_url, data)
+
+        for error_message in error_messages:
+            self.assertContains(response, error_message)
+
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:196]: This scenario is covering the case where the main form is invalid, which isn't handled directly by our implementation so we don't need this. For completeness this is handled by Django's CreateView (FormView in reality) and we trust that Django is doing the right thing here in terms of displaying errors.

However we're interested in the inline form called notify_formset in the implementation. There's no test to validate whether or not we're handling the case when this becomes invalid. See testplans/tests/test_views.py around line 70 for clues.

File: tcms/testcases/tests/test_views.py [hunk 2]
```diff
@@ -179,6 +179,21 @@ def setUpTestData(cls):
         user_should_have_perm(cls.tester, 'testcases.add_testcase')
         user_should_have_perm(cls.tester, 'testcases.view_testcase')
 
+    def test_form_invalid(self):
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:182]: There also doesn't seem to be a descendant of PermissionsTestCase for this view.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testcases/tests/test_views.py (hunk 1 above)
Comment (rish07) [reply tcms/testcases/tests/test_views.py:207]: @atodorov I have made the changes. Please let me know if they are okay.

File: tcms/testcases/tests/test_views.py [hunk 3]
```diff
@@ -179,6 +179,32 @@ def setUpTestData(cls):
...
+
+        response = self.client.post(self.new_case_url, data)
+
+        self.assertContains(response, _('New TestCase'))
+        self.assertContains(
+            response,
+            '<input class="bootstrap-switch" name="email_settings-0-auto_to_case_tester" '
+            'type="checkbox"', html=False)
+
+    def test_new_test_case_without_permissions(self):
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:196]: Must descend from PermissionsTestCase, see how it is used in other parts of the test suite.

File: tcms/testcases/tests/test_views.py [hunk 4]
```diff
@@ -179,6 +179,32 @@ def setUpTestData(cls):
         user_should_have_perm(cls.tester, 'testcases.add_testcase')
         user_should_have_perm(cls.tester, 'testcases.view_testcase')
 
+    def test_notify_formset_invalid(self):
+
+        data = self.data.copy()
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:184]: The place where you have copied this from contains an important comment as to why we are doing it this way. Copy the comment here as well b/c it is not immediately obvious why this test works like that.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testcases/tests/test_views.py (hunk 3 above)
Comment (rish07) [reply tcms/testcases/tests/test_views.py]: I removed this completely cause this edge case was already being tested under test_create_test_case_without_permissions.

File: tcms/testcases/tests/test_views.py (hunk 4 above)
Comment (rish07) [reply tcms/testcases/tests/test_views.py:185]: Done!

File: tcms/testcases/tests/test_views.py (hunk 3 above)
Comment (atodorov) [reply tcms/testcases/tests/test_views.py]: I don't understand. You say you have removed the scenario "test_new_test_case_without_permissions" but the diff section shows that nothing has been removed. OTOH "test_create_test_case_without_permissions" does exist but it si very old and needs to be updated to inherit from PermissionsTestCase. 

I'm not sure if the PR was old or if something else is going on here.  Can you try rebasing onto the latest master branch so I can investigate.

Comment (atodorov) [reply tcms/testcases/tests/test_views.py]: I don't understand. You say you have removed the scenario "test_new_test_case_without_permissions" but the diff section shows that nothing has been removed. OTOH "test_create_test_case_without_permissions" does exist but it si very old and needs to be updated to inherit from PermissionsTestCase. 

I'm not sure if the PR was old or if something else is going on here.  Can you try rebasing onto the latest master branch so I can investigate.

Comment (rish07) [reply tcms/testcases/tests/test_views.py]: Can you please check the latest code? I have created a new class TestNewCasePermission that inherits from PermissionsTestCase.

File: tcms/testcases/tests/test_views.py [hunk 5]
```diff
@@ -228,6 +232,95 @@ def _assertTestCase(self, test_case):
...
+    def verify_get_with_permission(self):
+        response = self.client.get(self.url)
+        self.assertEqual(response.status_code, 200)
+
+    def verify_get_without_permission(self):
+        response = self.client.get(self.url)
+        redirect_url = "{0}?next={1}".format(
+            reverse('tcms-login'), reverse('testcases-new')
+        )
+        self.assertRedirects(response, redirect_url)
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:305]: verify get/post without permission should already be handled by PermissionsTestCase itself and we need to defined these methods in very race circumstances. Try commenting them out and see if the test works. If it does then you don't need them explicitly defined.

File: tcms/testcases/tests/test_views.py [hunk 6]
```diff
@@ -228,6 +232,95 @@ def _assertTestCase(self, test_case):
...
+            'email_settings-INITIAL_FORMS': '1',
+            'email_settings-MIN_NUM_FORMS': '0',
+            'email_settings-MAX_NUM_FORMS': '1',
+        }
+
+        super().setUpTestData()
+
+    def verify_get_with_permission(self):
+        response = self.client.get(self.url)
+        self.assertEqual(response.status_code, 200)
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:298]: this needs to assert that we are on the correct page - for example by checking a well known headline, field/set of fields or the page title.

File: tcms/testcases/tests/test_views.py [hunk 7]
```diff
@@ -228,6 +232,95 @@ def _assertTestCase(self, test_case):
...
+        redirect_url = "{0}?next={1}".format(
+            reverse('tcms-login'), reverse('testcases-new')
+        )
+        self.assertRedirects(response, redirect_url)
+
+    def verify_post_with_permission(self):
+        response = self.client.post(self.url, self.post_data)
+        test_case = TestCase.objects.get(summary=self.summary)
+        redirect_url = reverse('testcases-get', args=[test_case.pk])
+        self.assertRedirects(response, redirect_url, status_code=302, target_status_code=302)
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:311]: you can post with `follow=True` which will follow all redirects and then you can assert that the response is the page displaying a test case with the expected values. 

Checking some well known elements/fields on the page is a good idea.

File: tcms/testcases/tests/test_views.py [hunk 8]
```diff
@@ -228,6 +232,95 @@ def _assertTestCase(self, test_case):
...
+        cls.summary = 'summary'
+
+        cls.post_data = {
+            'author': cls.tester.pk,
+            'summary': cls.summary,
+            'default_tester': cls.tester.pk,
+            'product': cls.case.category.product.pk,
+            'category': cls.case.category.pk,
+            'case_status': cls.case_status_confirmed.pk,
+            'priority': cls.case.priority.pk,
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:270]: The values for most of these fields are not used anywhere outside of the setup method so you don't need them as class attributes.

Comment (atodorov) [inline tcms/testcases/tests/test_views.py:270]: The values for most of these fields are not used anywhere outside of the setup method so you don't need them as class attributes. They can be hard-coded as well.

File: tcms/testcases/tests/test_views.py [hunk 9]
```diff
@@ -228,6 +232,95 @@ def _assertTestCase(self, test_case):
...
+            product=cls.product,
+            product_version=cls.version)
+
+        cls.case = TestCaseFactory(
+            author=cls.tester,
+            default_tester=None,
+            reviewer=cls.tester,
+            case_status=cls.case_status_confirmed,
+            plan=[cls.plan])
+        cls.case.save()  # will generate history object
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:260]: You do need some of the values for plan, category, etc but don't need to create yet another test case object in the database because it will be created when the data is POSTed successfully.

File: tcms/testcases/tests/test_views.py [hunk 10]
```diff
@@ -228,6 +232,95 @@ def _assertTestCase(self, test_case):
...
+class TestNewCasePermission(PermissionsTestCase):
+    permission_label = 'testcases.add_testcase'
+    http_method_names = ['get', 'post']
+    url = reverse('testcases-new')
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.tester = UserFactory()
+        cls.case_status_confirmed = TestCaseStatus.objects.get(name='CONFIRMED')
+        cls.case_status_proposed = TestCaseStatus.objects.get(name='PROPOSED')
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:244]: Use methods get_confirmed/get_proposed in the TestCaseStatus class and make sure you are only creating data which is necessary. If it is not used then it's not needed.

File: tcms/testcases/tests/test_views.py [hunk 11]
```diff
@@ -228,6 +232,95 @@ def _assertTestCase(self, test_case):
...
 
 
+class TestNewCasePermission(PermissionsTestCase):
+    permission_label = 'testcases.add_testcase'
+    http_method_names = ['get', 'post']
+    url = reverse('testcases-new')
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.tester = UserFactory()
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:242]: Already part of LoggedInTestCase from which PermissionsTestCase inherits.

**Review by atodorov — CHANGES_REQUESTED** [review]

**Comment by rish07** [comment]
@atodorov fixed all the changes. Please let me know if there are any more changes

**Review by atodorov — CHANGES_REQUESTED** [review]
Assertions look much better but there are still unresolved comments from previous review.

**Comment by rish07** [comment]
@atodorov can you mark the ones that are complete as resolved so I know which ones to work on then.

File: tcms/testcases/tests/test_views.py (hunk 8 above)
Comment (rish07) [reply tcms/testcases/tests/test_views.py]: We do need case as a class attribute because I am using self.case under _assertTestCase in TestNewCasePermission

File: tcms/testcases/tests/test_views.py (hunk 9 above)
Comment (rish07) [reply tcms/testcases/tests/test_views.py]: I have removed cls.plan and have created a case. If I don't create a case object from TestCaseFactory then there won't be any entry in the database to fetch later to assert in verify_post_with_permission

File: tcms/testcases/tests/test_views.py (hunk 10 above)
Comment (rish07) [reply tcms/testcases/tests/test_views.py]: I do need the 'case_status' parameter in cls.post_data while POSTing because it is a required field. And for that I need cls.case_status_confirmed. I have removed cls.case_status_proposed

File: tcms/testcases/tests/test_views.py [hunk 12]
```diff
@@ -228,6 +232,82 @@ def _assertTestCase(self, test_case):
...
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, _('New TestCase'))
+
+    def verify_post_with_permission(self):
+        response = self.client.post(self.url, self.post_data, follow=True)
+        test_case = TestCase.objects.get(summary=self.summary)
+        redirect_url = reverse('testcases-get', args=[test_case.pk])
+
+        self.assertRedirects(response, redirect_url, status_code=302, target_status_code=200)
+        self._assertTestCase(test_case)
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:297]: This method is overkill here. Line 293 above makes sure the newly created record is in the database and then the assertRedirects verifies that the browser was redirected correctly. At most you can check that the summary is displayed on the new page (isn't strictly necessary). The rest basically validates the same.

File: tcms/testcases/tests/test_views.py (hunk 10 above)
Comment (atodorov) [reply tcms/testcases/tests/test_views.py]: Not using get_confirmed method like shown previously.

**Review by atodorov — CHANGES_REQUESTED** [review]
need changes

File: tcms/testcases/tests/test_views.py (hunk 12 above)
Comment (rish07) [reply tcms/testcases/tests/test_views.py]: Removed the unnecessary part.

**Comment by rish07** [comment]
@atodorov Is this fine?

File: tcms/testcases/tests/test_views.py [hunk 13]
```diff
@@ -228,6 +232,71 @@ def _assertTestCase(self, test_case):
...
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, _('New TestCase'))
+
+    def verify_post_with_permission(self):
+        response = self.client.post(self.url, self.post_data, follow=True)
+        test_case = TestCase.objects.get(summary=self.summary)
+        redirect_url = reverse('testcases-get', args=[test_case.pk])
+
+        self.assertRedirects(response, redirect_url, status_code=302, target_status_code=200)
+        self.assertEqual(test_case.summary, self.summary)
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:297]: This is equivalent to line 293. Search for self.summary or test_case.summary inside the response.

File: tcms/testcases/tests/test_views.py [hunk 14]
```diff
@@ -3,20 +3,20 @@
...
-from tcms.testcases.models import TestCase, TestCasePlan
+from tcms.testcases.models import TestCase, TestCasePlan, TestCaseStatus
 from tcms.testcases.views import get_selected_testcases
 from tcms.testruns.models import TestExecutionStatus
 from tcms.tests import (BaseCaseRun, BasePlanCase, remove_perm_from_user,
-                        user_should_have_perm)
-from tcms.tests.factories import LinkReferenceFactory, TestCaseFactory
+                        user_should_have_perm, PermissionsTestCase)
+from tcms.tests.factories import (LinkReferenceFactory, TestCaseFactory,
+                                  UserFactory, ProductFactory, TestPlanFactory, VersionFactory)
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:19]: There are unused imports, see flake8 results. If the line becomes too long you can also use multiple imports on separate lines.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testcases/tests/test_views.py (hunk 13 above)
Comment (rish07) [reply tcms/testcases/tests/test_views.py:297]: Done

File: tcms/testcases/tests/test_views.py (hunk 14 above)
Comment (rish07) [reply tcms/testcases/tests/test_views.py]: Done @atodorov

File: tcms/testcases/tests/test_views.py [hunk 15]
```diff
@@ -228,6 +231,71 @@ def _assertTestCase(self, test_case):
...
+
+        case = TestCaseFactory(
+            author=cls.tester,
+            default_tester=None,
+            reviewer=cls.tester,
+            case_status=case_status_confirmed,
+        )
+        case.save()  # creates a new case entry in the database
+
+        cls.post_data.update({
```
Comment (atodorov) [inline tcms/testcases/tests/test_views.py:258]: This makes pylint unhappy but is a false positive in this case. There is a line `# pylint: disable=objects-update-used` in TestEditCase which you can move to the top of this file, together with the other disables and it will silence pylint.

**Review by atodorov — CHANGES_REQUESTED** [review]
Looks good, need to make pylint happy.

File: tcms/testcases/tests/test_views.py (hunk 15 above)
Comment (rish07) [reply tcms/testcases/tests/test_views.py:258]: done @atodorov

**Comment by rish07** [comment]
Anything else @atodorov ?
